### PR TITLE
[#49] [#50] [Integrate] As a user, I can see and answer a text-area question

### LINKED ISSUE #49: [Integrate] As a user, I can see a text-area question
## Why
A text-area question type is a question type that allows a user to fill in a multi-line text input. It is commonly used when we would like to ask an open-end question, such as asking for suggestions or comments.

**Example questions**
- _Do you have any suggestions?_
- _Your comments are appreciated!_

## Acceptance Criteria
When a Question model has `display_type` as `textarea`,
- Display a text-area question
- Use the question's short text as the textbox's placeholder

## Design
[Text-area Question](https://www.figma.com/file/GjRPOjDyZ6f4EDL3wKarRK/Challenge-Mobile-App?node-id=2632%3A674)

### LINKED ISSUE #50: [Integrate] As a user, I can answer a text-area question
## Why
A valid text-area question must have one answer model in its answers array. We will always use the first answer of the array and attach the user's text input to it when storing the user's answer for a text-area question.

## Acceptance Criteria
When tapping the Next button (or the Submit button) on a text-area question, store the user's answer,
```
{
  "id": "{{question-id}}",
  "answers": [
    {
      "id": "{{first-answer-id}}",
      "answer": "{{text-input}}"
    }
  ]
}
```

## Design
[Text-area Question](https://www.figma.com/file/GjRPOjDyZ6f4EDL3wKarRK/Challenge-Mobile-App?node-id=2632%3A674)

### PR DESCRIPTION
**Pull request by kaungkhantsoe**
Closes #49 #50 

## What happened 👀

A text-area question type is a question type that allows a user to fill in a multi-line text input. It is commonly used when we would like to ask an open-end question, such as asking for suggestions or comments.

**Example questions**
- _Do you have any suggestions?_
- _Your comments are appreciated!_

## Insight 📝

When a Question model has `display_type` as `textarea`,
- Display a text-area question
- Use the question's short text as the textbox's placeholder
- Add new variable named `answer` in `SurveyAnswer`
- Fixed parsing wrong answer in `SurveyAnswer.toSurveyAnswerRequest`

## Proof Of Work 📹

https://user-images.githubusercontent.com/32578035/196900948-5679460b-03db-43be-99d8-b652d4b2b1b3.mp4

### CODE REVIEW

File: app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt [hunk 1]
```diff
@@ -28,12 +27,13 @@ fun SurveyTextAreaQuestionScreen(
...
     onChooseAnswer: (answers: List<SurveyAnswer>) -> Unit
 ) {
-    var answer by remember { mutableStateOf("") }
+    var answer by remember { mutableStateOf(answers.first().answer) }
 
     TextField(
         value = answer,
         onValueChange = {
             answer = it
+            onChooseAnswer(answers.map { surveyAnswer -> surveyAnswer.copy(answer = answer, selected = true) })
```
Comment (ryan-conway) [inline app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt]: Shall we have some sort of debounce on this? Looks like it will update the question on every keystroke. What do you think? 😄

**Review by ryan-conway — COMMENTED** [review]
Only one concern aside from the failing tests 🙏🏻

File: app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt [hunk 2]
```diff
@@ -9,32 +9,45 @@ import androidx.compose.runtime.Composable
...
 fun SurveyTextAreaQuestionScreen(
     answers: List<SurveyAnswer>,
-    shortText: String,
-    onChooseAnswer: (answers: List<SurveyAnswer>) -> Unit
+    onAnswerChange: (answers: List<SurveyAnswer>) -> Unit
 ) {
-    var answer by remember { mutableStateOf("") }
+    var answer by remember { mutableStateOf(answers.first().answer) }
+    val coroutineScope = rememberCoroutineScope()
+    var debounceJob: Job? = null
```
Comment (ryan-conway) [inline app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt]: We can make use of `LaunchedEffect` here so that we don't have to manually manage our job 😄 
```suggestion
    LaunchedEffect(key1 = answer) {
        delay()
        onAnswerChange()
    }
```

File: app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt [hunk 3]
```diff
@@ -9,32 +9,45 @@ import androidx.compose.runtime.Composable
...
         value = answer,
-        onValueChange = { answer = it },
+        onValueChange = {
+            answer = it
+            if (debounceJob?.isActive == true) debounceJob?.cancel()
+            debounceJob = coroutineScope.launch {
+                delay(DELAY_VALUE_CHANE_MILLI)
+                onAnswerChange(answers.map { surveyAnswer -> surveyAnswer.copy(answer = answer, selected = true) })
+            }
+        },
```
Comment (ryan-conway) [inline app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt]: Then we can revert this change 😄 
```suggestion
        onValueChange = { answer = it } 
```

File: app/src/main/java/com/kks/nimblesurveyjetpackcompose/model/SurveyAnswer.kt [hunk 4]
```diff
@@ -4,12 +4,13 @@ import com.kks.nimblesurveyjetpackcompose.model.request.SurveyAnswerRequest
 
 data class SurveyAnswer(
     val id: String,
-    var text: String,
+    val text: String,
     val displayOrder: Int,
-    var selected: Boolean = false
+    var selected: Boolean = false,
+    var answer: String
```
Comment (ryan-conway) [inline app/src/main/java/com/kks/nimblesurveyjetpackcompose/model/SurveyAnswer.kt]: if we set a default value for `answer` we don't need to update `IncludedResponse.toSurveyAnswer()` and the previews anymore 😉 
```suggestion
    var answer: String = ""
```

File: app/src/main/java/com/kks/nimblesurveyjetpackcompose/model/response/IncludedResponse.kt [hunk 5]
```diff
@@ -108,5 +108,6 @@ fun IncludedAnswerResponse.toSurveyAnswer() =
     SurveyAnswer(
         id = id.orEmpty(),
         text = attributes?.text.orEmpty().trim(),
-        displayOrder = attributes?.displayOrder ?: 0
+        displayOrder = attributes?.displayOrder ?: 0,
+        answer = ""
```
Comment (ryan-conway) [inline app/src/main/java/com/kks/nimblesurveyjetpackcompose/model/response/IncludedResponse.kt]:
```suggestion
        displayOrder = attributes?.displayOrder ?: 0
```

File: app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt [hunk 6]
```diff
@@ -49,16 +62,13 @@ fun SurveyTextAreaQuestionScreen(
...
-            }
     )
 }
 
 @Preview(showBackground = true, backgroundColor = 0)
 @Composable
 fun SurveyTextAreaQuestionScreenPreview() {
-    SurveyTextAreaQuestionScreen(emptyList(), "", {})
+    SurveyTextAreaQuestionScreen(
+        listOf(SurveyAnswer("", "Your thought", 0, false, ""))
```
Comment (ryan-conway) [inline app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt]:
```suggestion
        listOf(SurveyAnswer("", "Your thought", 0, false))
```

File: app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyChoiceQuestionScreen.kt [hunk 7]
```diff
@@ -115,8 +115,8 @@ fun SurveyChoiceQuestionScreen(
 fun PreviewSurveyChoiceQuestionScreen() {
     SurveyChoiceQuestionScreen(
         answers = listOf(
-            SurveyAnswer("", "Choice 1", 0, false),
-            SurveyAnswer("", "Choice 2", 1, true)
+            SurveyAnswer("", "Choice 1", 0, false, ""),
+            SurveyAnswer("", "Choice 2", 1, true, "")
```
Comment (ryan-conway) [inline app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyChoiceQuestionScreen.kt]:
```suggestion
            SurveyAnswer("", "Choice 1", 0, false),
            SurveyAnswer("", "Choice 2", 1, true)
```

**Review by ryan-conway — COMMENTED** [review]
A few comments this time, the rest LGTM 👍🏻

**Review by ryan-conway — APPROVED** [review]
Nice, LGTM 👍🏻

File: app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt [hunk 8]
```diff
@@ -6,35 +6,42 @@ import androidx.compose.foundation.shape.RoundedCornerShape
...
 import androidx.compose.ui.graphics.Color
 import androidx.compose.ui.tooling.preview.Preview
 import androidx.compose.ui.unit.dp
 import androidx.compose.ui.unit.sp
 import com.kks.nimblesurveyjetpackcompose.model.SurveyAnswer
 import com.kks.nimblesurveyjetpackcompose.ui.presentation.common.SurveyText
 import com.kks.nimblesurveyjetpackcompose.ui.theme.White20
+import kotlinx.coroutines.delay
+
+private const val DELAY_VALUE_CHANE_MILLI = 300L
```
Comment (AVI5HEK) [inline app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt]: You mean `DELAY_VALUE_CHANGE_MILLI`?

**Review by AVI5HEK — COMMENTED** [review]
The rest lgtm.

File: app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt [hunk 9]
```diff
@@ -6,35 +6,42 @@ import androidx.compose.foundation.shape.RoundedCornerShape
...
+    }
 
     TextField(
         value = answer,
         onValueChange = { answer = it },
         shape = RoundedCornerShape(10.dp),
         placeholder = {
             SurveyText(
-                text = shortText,
+                text = answers.first().text,
```
Comment (AVI5HEK) [inline app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/survey/SurveyTextAreaQuestionScreen.kt]: What if once in a blue moon `answers` doesn't have any element?
```suggestion
                text = answers.firstOrNull()?.text.orEmpty(),
```

**Review by AVI5HEK — COMMENTED** [review]
Just one concern. The rest lgtm.

**Review by AVI5HEK — APPROVED** [review]
